src/transformStateWithClones.js

```diff
@@ -6,8 +6,51 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(state = {}, actions = []) {
+  const clonedState = JSON.parse(JSON.stringify(state));
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(clonedState, action.extraData, stateHistory);
+        break;
+      case 'removeProperties':
+        removeProperties(clonedState, action.keysToRemove, stateHistory);
+        break;
+      case 'clear':
+        clearProperties(clonedState, stateHistory);
+        break;
+    }
+  }
+}
+
+function addProperties(clonedState, extraData, stateHistory) {
+  Object.assign(clonedState, extraData);
+
+  const addHistory = { ...clonedState };
```

Comment:
The variable name 'addHistory' is not descriptive of its purpose. Consider renaming it to 'stateAfterAddition' to more accurately reflect its role.